hub/core/chunk_engine/generator.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+from typing import Generator, Tuple
+
+from hub.core.chunk_engine.exceptions import ChunkGeneratorError
+
+
+def chunk(
+    content_bytes: bytes, previous_num_bytes: int, chunk_size: int
+) -> Generator[Tuple[bytes, int], None, None]:
+    """
+    Generator function that chunks bytes.
+
+    Chunking is the process of taking the input `content_bytes` & breaking it up into smaller chunks,
+    the sizes of which are <= `chunk_size`.
+
+    Example 1:
+        content_bytes = b"101010101010"
+        chunk_size = 4
+        output_chunks = [b"1010", b"1010", b"1010"]
+
+        This is considered a "perfect fit" because all bytes fit perfectly into 3 chunks.
+
+
+    Example 2:
+        content_bytes = b"10101010111122"
+        chunk_size = 4
+        output_chunks = [b"1010", b"1010", b"1111", b"22"]
+
+        As you can see, the last chunk in `output_chunks` only has 2 bytes (chunk_size=4), which means
+        this is not a perfect fit, but rather it has a partial chunk at the end.
+
+
+    Args:
+        content_bytes(bytes): Bytes object with the data to be chunked.
+        previous_num_bytes(int): The number of bytes in the previous chunk. This is helpful for filling the previous chunk before starting a new one.
+        chunk_size(int): Each individual chunk will be assigned this many bytes maximum.
+
+    Yields:
+        Each yield is a chunk of the `content_bytes`. Each chunk is of length (0, `chunk_size`].
+
+    Raises:
+        ChunkGeneratorError: If leftover bytes are negative or the previous chunk was invalid.
+    """
+
+    if len(content_bytes) <= 0:
+        return
+
+    if previous_num_bytes is None:
+        bytes_left = 0
+    else:
+        bytes_left = chunk_size - previous_num_bytes
+
+    content_num_bytes = len(content_bytes)
+
+    if bytes_left < 0:
+        raise ChunkGeneratorError(
```

Comment:
Not a fan of this error. Doesn't tell the user what went wrong from their side. They won't really be playing around with the chunking code I suppose, so encountering this error would be certainly a deadend.

Reply:
good point, you are right that errors should be useful to the user. i was looking at these core errors more from internal developer usage. i will update these messages.

Reply:
i actually think this error exactly describes the problem. anything that calls the `chunk(...)` method (which is located in the chunk engine, so theoretically the only thing calling this method should be the `ChunkManager`) should handle this error. 

this error will only happen if the `ChunkManager` messes up & tries to use an inconsistent `chunk_size`.

Reply:
so the solution to this is to translate this error to the user if it happens in the `ChunkManager`. for example, this may be a versioning issue or a corrupted meta file.

the `chunk(...)` function shouldn't have knowledge about any of this. but the `ChunkManager` should.